### ISSUE TITLE
NNMemoryBank not working with DataParallel

### ISSUE THREAD

**Issue by kfallah** [lightly-ai/lightly#968]
I have been using the NNMemoryBank as a component in my module and noticed that at each forward pass `NNMemoryBank.bank` is equal to `None`. This only occurs when my module is wrapped in `DataParallel`. As a result, throughout training my NN pairs are always random noise (surprisingly, this only hurt the contrastive learning performance by a few percentage point on linear probing??).

Here is a simple test case that highlights the issue:
```
import torch
from lightly.models.modules import NNMemoryBankModule
memory_bank = NNMemoryBankModule(size=1000)
print(memory_bank.bank)
memory_bank(torch.randn((100, 512)))
print(memory_bank.bank)

memory_bank = NNMemoryBankModule(size=1000)
memory_bank = torch.nn.DataParallel(memory_bank, device_ids=[0,1])
print(memory_bank.module.bank)
memory_bank(torch.randn((100, 512)))
print(memory_bank.module.bank)
```

The output of the first is `None` and a `torch.Tensor`, as expected. The output for the second is `None` for both.

**Comment by kfallah** [comment]
I just wanted to check back and see if there was any progress on this issue 🙂. Could this be fixed by making `bank` a parameter with `requires_grad=False` in `lightly.loss.memory_bank.MemoryBankModule`? I think this same issue may also occur with loading/saving the module, since `bank` is just a tensor.

**Comment by kfallah** [comment]
Note that this issue is a known limitation in the DataParallel, where the suggested solution is an in-place edit of a buffer or parameter (second warning blurb as of Nov 2022):
https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html 

Seems like the best design pattern would be to register `bank` as a buffer and do in-place edits in `forward`:
https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html
https://discuss.pytorch.org/t/what-is-the-difference-between-register-buffer-and-register-parameter-of-nn-module/32723

But also seems like there was a deliberate decision not to do this:
https://github.com/lightly-ai/lightly/blob/master/lightly/loss/memory_bank.py#L62
Would an extra `66536 x 256 = ~16 MB` really make a big difference (`65536` is the memory bank size from NNCLR, `256` for the projected feature dimension)?

**Comment by guarin** [comment]
Hi @kfallah,

Thanks a lot for investigating this! I think you are right and we should register the memory bank as a buffer. `register_buffer` takes an additional `persistent` argument which controls whether the buffer should be part of the state dict or not. I believe we can just set `persistent=False` to avoid saving the memory bank in the checkpoint.

Would you be interested in making a PR for this?

**Comment by kfallah** [comment]
I do not think I will have the time to make the PR, I would be deeply grateful if someone else had the chance.

**Comment by guarin** [comment]
I'll have a look as soon as I get some time.

**Comment by guarin** [comment]
@kfallah I made a PR with the change, this can hopefully be merged soon.
